refactor: use external_id instead of id for all admin routes

nyomanjyotisa opened the pull request:
Resolves https://github.com/antiwork/gumroad/issues/2070

# Summary
Migrated all admin routes to use `external_id` instead of numeric database IDs for better security, consistency, and public shareability. Updated controllers to fetch records by `external_id`, modified presenters to return `external_id` in JSON responses, and added backward-compatible redirects from numeric ID URLs to the external ID URLs

# After

## Users
### User show links updated to external_id, and accessing user show page with id redirects to external_id

https://github.com/user-attachments/assets/5c2a4fe9-ff88-4ee9-8da9-2608ef8a5554

https://gumroad.dev/admin/users/1 redirected to https://gumroad.dev/admin/users/4797619243095

## Products
### Product show links updated to external_id, and accessing product show page with id redirects to external_id

https://github.com/user-attachments/assets/1a6dfa1f-80b5-42ca-954b-272a47d110a9

https://gumroad.dev/admin/products/1 redirected to https://gumroad.dev/admin/products/fEGTaEpuKDsnDvf_MfecTA==

## Purchases
### Purchase show links updated to external_id, and accessing purchase show page with id redirects to external_id

https://github.com/user-attachments/assets/5d0c3c1f-63db-4968-8970-981b1f449bdc

https://gumroad.dev/admin/purchases/1 redirected to https://gumroad.dev/admin/purchases/fEGTaEpuKDsnDvf_MfecTA==

## Affiliates

https://github.com/user-attachments/assets/08859988-44fe-490f-a522-2f75e1c18758

https://gumroad.dev/admin/affiliates/6 redirected to https://gumroad.dev/admin/affiliates/2944786465943

## Merchant Accounts

https://github.com/user-attachments/assets/ada1e7fc-084a-49ed-97e9-eff922d926b8

https://gumroad.dev/admin/merchant_accounts/4 redirected to https://gumroad.dev/admin/merchant_accounts/qXyyQQfPkWq7Mt9c4ot2YA==

## Payouts

https://github.com/user-attachments/assets/d9fc1fa0-0ef5-4594-8728-8dce1a24a8eb

https://gumroad.dev/admin/payouts/1 redirected to https://gumroad.dev/admin/payouts/fEGTaEpuKDsnDvf_MfecTA==

# Test Results
specs executed: 
```
bundle exec rspec \
  spec/requests/admin \
  spec/presenters/admin \
  spec/controllers/admin
```

<img width="515" height="77" alt="image" src="https://github.com/user-attachments/assets/07ff928d-3490-4c2f-b2c7-7845e98080e7" />

<img width="575" height="126" alt="image" src="https://github.com/user-attachments/assets/7f6e83ad-287c-461f-87cd-a938da16c665" />



# AI Disclosure
Cursor - Claude Sonnet 4.5 was used to help me identify some edge cases that I might missed on refactoring the code, and also help on updating the specs

# Confirmation
I have watched the Gumroad PR reviews live streaming video

nyomanjyotisa commented on app/controllers/admin/purchases_controller.rb:
```diff
@@ -2,7 +2,7 @@
 class Admin::PurchasesController < Admin::BaseController
   before_action :fetch_purchase, only: %i[cancel_subscription refund refund_for_fraud refund_taxes_only resend_receipt
-                                          show sync_status_with_charge_processor block_buyer unblock_buyer undelete]
+                                          show sync_status_with_charge_processor block_buyer unblock_buyer undelete update_giftee_email]
```
Added `update_giftee_email` here so we can use `@purchase` inside `update_giftee_email` function. Since `params[:id]` no longer exists, and has been replaced by `params[:external_id]`

nyomanjyotisa reviewed (commented):
Hi @EmCousin Could help to review this PR? I have addressed your feedbacks from [this PR](https://github.com/antiwork/gumroad/pull/2147)

For the redirections, we should keep it based on [this comment](https://github.com/antiwork/gumroad/pull/2147#discussion_r2582932549)

cc @ershad

EmCousin commented on app/controllers/admin/affiliates/base_controller.rb:
```diff
@@ -5,6 +5,6 @@ class Admin::Affiliates::BaseController < Admin::BaseController
   private
     def set_affiliate_user
-      @affiliate_user = User.find(params[:affiliate_id])
+      @affiliate_user = User.find_by_external_id(params[:affiliate_external_id])
```
Same [comment](https://github.com/antiwork/gumroad/pull/2204/files#r2588942794)

EmCousin commented on app/controllers/admin/users/guids_controller.rb:
```diff
@@ -2,10 +2,19 @@
+
+    guids_to_users = events_by_guid.map do |browser_guid, events|
+      { guid: browser_guid, user_external_ids: events.map { |e| external_ids_by_user_id[e.user_id] } }
+    end
```
This could be simplified by having a `Event.belongs_to :user, optional: true` relationship, as well as `Event.delegate :external_id, to: :user, prefix: true, allow_nil: true` 

```ruby
class Event < ApplicationRecord
  # [...]
  belongs_to :user, optional: true,
  delegate :external_id, to: :user, prefix: true, allow_nil: true
  # [...]
end
```
Then this could be similar to the old code:
```ruby
user = User.find_by_external_id(user_param) || e404
guids = Event.where(user_id: user.id).distinct.pluck(:browser_guid)
guids_to_users = Event.select(:user_id, :browser_guid).by_browser_guid(guids)
                      .where.not(user_id: nil).distinct
                      .includes(:user)
                      .group_by(&:browser_guid)
                      .map { |browser_guid, events| { guid: browser_guid, user_ids: events.map(&:user_external_id) } }
```

nyomanjyotisa commented:
@EmCousin I have addressed all your feedback above. Could you please take another look?

Some UI changes (ID text to External ID):
| Before | After |
|-----------|-----------|
|  <img width="1470" height="795" alt="image" src="https://github.com/user-attachments/assets/710a1dc6-965c-40da-8c6f-3a926120a1f2" />  | <img width="1470" height="795" alt="image" src="https://github.com/user-attachments/assets/7cd387b2-ea04-42a0-a392-20ed054601a8" /> |
|  <img width="1470" height="797" alt="image" src="https://github.com/user-attachments/assets/2e6a4274-7164-417b-bf82-b51861ef1be6" />  |  <img width="1470" height="795" alt="image" src="https://github.com/user-attachments/assets/daeb1143-401e-4a6b-9fd8-8a60b7f7482b" /> |
|  <img width="1470" height="794" alt="image" src="https://github.com/user-attachments/assets/fd9e3701-4916-47e0-b6e7-e5bcbb0c0190" />  |  <img width="1470" height="795" alt="image" src="https://github.com/user-attachments/assets/30a4512a-5ffb-4a77-85e3-06fb4fcb2453" />  |
|  <img width="1470" height="794" alt="image" src="https://github.com/user-attachments/assets/34e60d1a-d648-44d3-8a5b-3fda2d3ee2f9" />   |  <img width="1470" height="795" alt="image" src="https://github.com/user-attachments/assets/30a4512a-5ffb-4a77-85e3-06fb4fcb2453" />  |
|  <img width="1470" height="792" alt="image" src="https://github.com/user-attachments/assets/e3bb495a-3a22-43cf-90ab-ab83aa962eb2" /> |  <img width="1470" height="787" alt="image" src="https://github.com/user-attachments/assets/e78ecfca-0e8e-4e54-a4a6-879c3c870f3c" />  |
|  <img width="1470" height="798" alt="image" src="https://github.com/user-attachments/assets/d0bb057f-3688-49e7-9f00-b3ce0c21daa3" />  |  <img width="1470" height="792" alt="image" src="https://github.com/user-attachments/assets/d77b4a57-eff7-45b0-a106-8400239f4020" />  |
